File: app/src/main/java/de/bikebean/app/db/settings/settings/NumberSetting.java
package de.bikebean.app.db.settings.settings;

import androidx.annotation.NonNull;

import java.util.List;

import de.bikebean.app.db.settings.Setting;
import de.bikebean.app.db.sms.Sms;
import de.bikebean.app.db.sms.SmsFactory;
import de.bikebean.app.db.state.State;
import de.bikebean.app.db.state.StateFactory;

public abstract class NumberSetting extends Setting {

    private final @NonNull List<? extends RawNumberSettings> list;
    private final @NonNull State numberState;

    protected NumberSetting(final @NonNull String wappString, final @NonNull Sms sms,
                            final @NonNull State.KEY key, final @NonNull State numberState,
                            final @NonNull List<? extends RawNumberSettings> list) {
        super(
                StateFactory.createStringState(
                        sms, key, wappString, State.STATUS.CONFIRMED
                ),
                Setting::addToList
        );

        this.numberState = numberState;
        this.list = list;
    }

    public NumberSetting(final @NonNull String wappString, final @NonNull State.KEY key,
                         final @NonNull State numberState,
                         final @NonNull List<? extends RawNumberSettings> list) {
        super(
                StateFactory.createStringState(
                        SmsFactory.createNullSms(), key,
                        wappString, State.STATUS.UNSET),
                Setting::addToList
        );

        this.numberState = numberState;
        this.list = list;
    }

    public final @NonNull List<? extends RawNumberSettings> getList() {
        return list;
    }

    public final int getNumber() {
        return getNumberState().getValue().intValue();
    }

    public final @NonNull State getNumberState() {
        return numberState;
    }

    public abstract static class RawNumberSettings {}
}

File: app/src/main/java/de/bikebean/app/db/settings/settings/add_to_list_settings/NumberSetting.kt
package de.bikebean.app.db.settings.settings.add_to_list_settings

import de.bikebean.app.db.settings.settings.AddToListSetting
import de.bikebean.app.db.sms.Sms
import de.bikebean.app.db.sms.SmsFactory
import de.bikebean.app.db.state.State
import de.bikebean.app.db.state.State.KEY
import de.bikebean.app.db.state.StateFactory

abstract class NumberSetting : AddToListSetting {

    val numberState: State

    protected constructor(
            wappString: String, sms: Sms, key: KEY,
            _numberState: State
    ) : super(
            StateFactory.createStringState(sms, key, wappString, State.STATUS.CONFIRMED)
    ) {
        numberState = _numberState
    }

    constructor(wappString: String, key: KEY,
                _numberState: State
    ) : super(
            StateFactory.createStringState(
                    SmsFactory.createNullSms(), key,
                    wappString, State.STATUS.UNSET)
    ) {
        numberState = _numberState
    }

    val number: Int
        get() = numberState.value.toInt()

    abstract class RawNumberSettings
    abstract val list: List<RawNumberSettings?>

    companion object {
        fun getNumberFromString(string: String): Int {
            return if (string.isEmpty()) 0 else string.split("\n").size
        }
    }
}